src/boot/boot.h

```diff
@@ -0,0 +1,46 @@
+#define physical_from_virtual(__x) u64_from_pointer(__x)
+
+typedef struct tagged_allocator {
+    struct heap h;
+    u8 tag;
+    heap parent;
+} *tagged_allocator;
+
+static inline u64 tagged_allocate(heap h, bytes length)
+{
+    tagged_allocator ta = (void *)h;
+    u64 base = allocate_u64(ta->parent, length + 1);
+    return base + 1;
+}
+
+static inline void tagged_deallocate(heap h, u64 a, bytes length)
+{
+    tagged_allocator ta = (void *)h;
+    deallocate_u64(ta->parent, a - 1, length + 1);
+}
+
+static inline heap allocate_tagged_region(heap h, u64 tag)
+{
+    tagged_allocator ta = allocate(h, sizeof(struct tagged_allocator));
+    ta->h.alloc = tagged_allocate;
+    ta->h.dealloc = tagged_deallocate;
+    ta->tag = tag;
+    ta->parent = h;
+    return (heap)ta;
+}
+
+static inline void *tag(void *v, u16 tval)
+{
+    *((u8 *)v-1) = tval;
+    return v;
+}
+
+static inline u16 tagof(void *v)
+{
+    return *((u8 *)v-1);
+}
```

Comment:
Same as above, should return a u8 instead?

Reply:
Returning a u8 now